packages/modals/src/elements/DrawerModal/DrawerModal.tsx

```diff
@@ -0,0 +1,249 @@
+  Header: typeof Header;
+  Body: typeof Body;
+  Close: typeof Close;
+  Footer: typeof StyledDrawerModalFooter;
+  FooterItem: typeof StyledDrawerModalFooterItem;
+}
+
+export interface IDrawerModalProps {
+  /**
+   * Determines whether the `DrawerModal` is opened or closed
+   */
+  isOpen?: boolean;
+  /**
+   * HTML attributes to spread onto backdrop element
+   */
+  backdropProps?: any;
+  /**
+   * Callback when a close action has been completed.
+   * Can be triggered from the backdrop and Close icon.
+   * @param {Object} event - DOM event that triggered the close action
+   */
+  onClose?: (event: TransitionEvent) => void;
+  /**
+   * The root ID to use for descendants. A unique ID is created if none is provided.
+   **/
+  id?: string;
+  /**
+   * The DOM element where the drawer modal will be rendered to
+   **/
+  appendToNode?: Element;
+  /**
+   * Determines whether to apply keyboard focus on the drawer modal on mount
+   */
+  focusOnMount?: boolean;
+  /**
+   * Determines whether to return keyboard focus to the element that triggered the drawer modal
+   */
+  restoreFocus?: boolean;
+  children?: any;
+}
+
+export const DrawerModal = forwardRef<
+  HTMLDivElement,
+  IDrawerModalProps & ThemeProps<DefaultTheme> & HTMLAttributes<HTMLDivElement>
+>(
+  (
+    { id, isOpen, onClose, backdropProps, appendToNode, focusOnMount, restoreFocus, ...props },
+    ref
+  ) => {
+    const modalRef = useRef<HTMLDivElement>(null);
+    const theme = useContext(ThemeContext);
+    const environment = useDocument(theme);
+    const { rtl } = theme;
+    /**
+     * The `showDrawer` local state is synced to `isOpen` prop. The DrawerModal component needs to control
+     * local state in order to know when (CSS transition ends) to unmount the DrawerModal component.
+     */
+    const [showDrawer, setShowDrawer] = useState(false);
+
+    useFocusVisible({ scope: modalRef, relativeDocument: environment });
+
+    const unmountDrawer = useCallback(
+      (event: TransitionEvent) => {
+        if ((event.target as HTMLDivElement).getAttribute('role') === 'dialog') {
+          setShowDrawer(false);
+          onClose && onClose(event);
+        }
+      },
+      [setShowDrawer, onClose]
+    );
+
+    const {
+      getTitleProps,
+      getCloseProps,
+      getContentProps,
+      getBackdropProps,
+      getModalProps
+    } = useModal({
+      id,
+      modalRef,
+      focusOnMount,
+      restoreFocus,
+      onClose: () => {
+        if (modalRef.current) {
+          const { width } = modalRef.current.getBoundingClientRect();
+          modalRef.current.addEventListener('transitionend', unmountDrawer);
+
+          if (rtl) {
+            modalRef.current.style.transform = 'translateX(0px)';
+          } else {
+            modalRef.current.style.right = `-${width}px`;
+            modalRef.current.style.transform = `translateX(${width}px)`;
+          }
+        }
+      }
```

Comment:
There is a lot logic here for the unmount logic. We currently don't animate the unmount of our other Modal components. @rossmoody are we sure that the fade-out is a large user benefit?

If so we may need to think about create a new type of API for this type of interaction. 

Some first thoughts are not basing the visibility on the `DrawerModal` mount/unmount cycle. We could include an `isVisible` prop which would allow the value to still be controlled via `onClose`, but helps us avoid the event listening. This could also allow us to start using react-transition-group.

But, my first hunch is to align with the other Modal animation patterns we have.

Reply:
I would say it's definitely a more pleasant user experience to witness the Drawer animate out as opposed to vanish.

Reply:
@austin94 with `react-transition-group` the unmount/mount logic can be removed 💯 . If the primary reason to kill the transition upon unmount is the logic, then it can keep it since `react-transition-group` will help remove that logic. 

@rossmoody I think a good reason to have the unmount transition for the `DrawerModal` is that the transition on mount is more emphasized than on a `Modal`. E.g. the transition duration is longer and the CSS property being transitioned is more apparent (`opacity` vs `translateX`). So transitioning on unmount for the `DrawerModal` makes it feel more natural.